internal/services/v1/permissions.go

```diff
@@ -535,86 +536,157 @@ func (ps *permissionServer) LookupSubjects(req *v1.LookupSubjectsRequest, resp v
+	}
+
+	if req.OptionalCursor != nil {
+		decodedCursor, err := cursor.DecodeToDispatchCursor(req.OptionalCursor, lsRequestHash)
+		if err != nil {
+			return ps.rewriteError(ctx, err)
 		}
+		currentCursor = decodedCursor
+	}
+
+	if req.OptionalConcreteLimit > 0 {
+		remainingConcreteLimit = int(req.OptionalConcreteLimit)
+	}
+
+	internalResponseCursor := &impl.DecodedCursor{
+		VersionOneof: &impl.DecodedCursor_V1{
+			V1: &impl.V1Cursor{
+				Revision:              atRevision.String(),
+				CallAndParametersHash: lsRequestHash,
+			},
+		},
+	}
+
+	for {
+		countSubjectsFound := 0
 
-		for _, foundSubject := range foundSubjects.FoundSubjects {
-			excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
+		ctxWithCancel, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		stream := dispatchpkg.NewHandlingDispatchStream(ctxWithCancel, func(result *dispatch.DispatchLookupSubjectsResponse) error {
+			foundSubjects, ok := result.FoundSubjectsByResourceId[req.Resource.ObjectId]
+			if !ok {
+				return fmt.Errorf("missing resource ID in returned LS")
 			}
 
-			excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				resolvedExcludedSubject, err := foundSubjectToResolvedSubject(ctx, excludedSubject, caveatContext, ds)
+			for _, foundSubject := range foundSubjects.FoundSubjects {
+				// Skip wildcards if requested they be skipped.
+				if req.WildcardOption == v1.LookupSubjectsRequest_WILDCARD_OPTION_EXCLUDE_WILDCARDS && foundSubject.SubjectId == tuple.PublicWildcard {
+					continue
+				}
+
+				excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
+				excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
+				for _, excludedSubject := range foundSubject.ExcludedSubjects {
+					excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
+
+					resolvedExcludedSubject, err := foundSubjectToResolvedSubject(ctx, excludedSubject, caveatContext, ds)
+					if err != nil {
+						return fmt.Errorf("error when resolving excluded subject: %w", err)
+					}
+
+					if resolvedExcludedSubject == nil {
+						continue
+					}
+
+					excludedSubjects = append(excludedSubjects, resolvedExcludedSubject)
+				}
+
+				subject, err := foundSubjectToResolvedSubject(ctx, foundSubject, caveatContext, ds)
+				if err != nil {
+					return fmt.Errorf("error when resolving subject: %w", err)
+				}
+				if subject == nil {
+					continue
+				}
+
+				// NOTE: we need to recompute the cursor here because we get multiple results back from DispatchLookupSubjects
+				// in one message.
+				dispatchCursor, err := graph.CursorForFoundSubjectID(subject.SubjectObjectId, result.AfterResponseCursor)
```

Comment:
out of scope, but for discussion in a GitHub issue: I think we should explore batching in the streaming APIs. The overhead of sending 1 element per message over the stream has overhead when the result list is large. Same rationale as what we've done with the bulk import / export APIs

Reply:
Yeah, but it makes cursoring MUCH harder; RR used to be batched and I had to un-batch it to get cursors working correctly